File: python/problems_claude/11.py
# Problem 11: Group Anagrams
# ** Difficulty: ** Medium
# **Time:** 20 minutes
# **Pattern:** Hash map with sorted key
# Group strings that are anagrams of each other.


def group_anagrams(strs):
    """
    Group strings that are anagrams of each other.

    Example:
    Input: ["eat", "tea", "tan", "ate", "nat", "bat"]
    Output: [["eat","tea","ate"], ["tan","nat"], ["bat"]]

    Note: Order of groups and order within groups may vary.
    """
    # Your code here
    hash_map = {}

    for s in strs:
        s_tmp = "".join(sorted(s.lower()))
        if s_tmp in hash_map:
            hash_map[s_tmp].append(s)
        else:
            hash_map[s_tmp] = [s]
    ret_list = []

    for key, value in hash_map.items():
        ret_list.append(value)

    return ret_list


# Test cases
result = group_anagrams(["eat", "tea", "tan", "ate", "nat", "bat"])
expected = [["eat", "tea", "ate"], ["tan", "nat"], ["bat"]]
# Convert to sets for comparison (order doesn't matter)
result_sets = [set(group) for group in result]
expected_sets = [set(group) for group in expected]
assert sorted(result_sets, key=str) == sorted(expected_sets, key=str)

result = group_anagrams([""])
assert result == [[""]]

result = group_anagrams(["a"])
assert result == [["a"]]

print("All tests passed!")

File: cpp/problems_2/11.cpp
// #### Problem 11: Move Zeroes
//
// **Difficulty:** Easy
// **Time:** 15 minutes
// **Pattern:** Two pointers technique
//
// Given an integer array `nums`, move all 0's to the end of it while
// maintaining the relative order of the non-zero elements.

#include <cassert>
#include <iostream>
#include <vector>

void moveZeroes(std::vector<int> &nums) {
  /*
  Move all zeros to end while maintaining order of non-zeros.

  Examples:
  Input: nums = [0,1,0,3,12]
  Output: [1,3,12,0,0]

  Input: nums = [0]
  Output: [0]
  */

  // Your code here
}

int main() {
  std::vector<int> nums1 = {0, 1, 0, 3, 12};
  moveZeroes(nums1);
  assert(nums1 == std::vector<int>({1, 3, 12, 0, 0}));

  std::vector<int> nums2 = {0};
  moveZeroes(nums2);
  assert(nums2 == std::vector<int>({0}));

  std::vector<int> nums3 = {1, 2, 3};
  moveZeroes(nums3);
  assert(nums3 == std::vector<int>({1, 2, 3}));

  std::cout << "Problem 11: All tests passed!" << std::endl;
  return 0;
}
